Set/ViewController.swift

```diff
@@ -8,12 +8,451 @@
+    }
+    
+    // Creates and returns a custom SetCardView for the provided setCard object.
+    private func getCardView(of card: SetCard) -> SetCardView {
+        // initiate "empty" cardView instance:
+        let viewCard = SetCardView(frame: CGRect())
+        
+        // Define a tap recognizer and assign it to every SetCardView
+        let tapRecognizer = UITapGestureRecognizer(target: self, action: #selector(cardTapHandler(recognizer:)))
+        viewCard.addGestureRecognizer(tapRecognizer)
+        
+        // set SetCardView color:
+        switch card.color {
+        case .typeOne: viewCard.color = .pink
+        case .typeTwo: viewCard.color = .purple
+        case .typeThree: viewCard.color = .green
+        }
+        // set SetCardView shapeType:
+        switch card.shapeType {
+        case .typeOne: viewCard.shape = .triangle
+        case .typeTwo: viewCard.shape = .circle
+        case .typeThree: viewCard.shape = .square
+        }
+        // set SetCardView numberOfShapes:
+        switch card.shapesNum {
+        case .one: viewCard.numberOfShapes = .single
+        case .two: viewCard.numberOfShapes = .double
+        case .three: viewCard.numberOfShapes = .triple
+        }
+        // set setCardView fillings
+        switch card.filling {
+        case .typeOne: viewCard.filling = .full
+        case .typeTwo: viewCard.filling = .outline
+        case .typeThree: viewCard.filling = .striped
+        }
+        return viewCard
+    }
+    
+    // Creates the mapper between setCard objects and their matching setCardViews
+    private func createCardsToCardsViewsMapper() {
+        
+        // When a SetGame instance is initialized, it has 69 cards in the deck and 12 cards in openCards:
+        for card in game.deck {
+            cardsToViewsMapper[card] = getCardView(of: card)
+        }
+        
+        for card in game.openCards {
+            cardsToViewsMapper[card] = getCardView(of: card)
+        }
+    }
+    
+    // Makes sure that the cardView of every open card in the game is a subview of openCardsCanvas
+    private func updateOpenCardsViews() {
+        
+        // remove from the canvas views of cards that were already matched (and aren't in openCards anymore)
+        for view in openCardsCanvas.subviews {
+            guard let index = cardsToViewsMapper.values.firstIndex(of: view) else {
+                print("Encountered an error.")
+                return
+            }
+            let card = cardsToViewsMapper.keys[index]
+            if !game.openCards.contains(card) {
+                view.removeFromSuperview()
+            }
+        }
+        
+        // adds views of cards that were recently popped from the deck to openCards
+        for card in game.openCards {
+            guard let currCardView = cardsToViewsMapper[card] else{
+                print("Couldn't find the view (cardView) for one of the open cards in the game")
```

Comment:
for this types of error it's better to use `assert()` since it will crash the app during development but not in "Release" mode and this is the sort of errors you want to make sure you're not missing if happens

Reply:
👍🏻👍🏻